File: sdks/java/src/test/java/io/delopay/sdk/ProvidersClientTest.java
package io.delopay.sdk;

import io.delopay.sdk.model.PaymentMethodsResponse;
import io.delopay.sdk.model.ProviderClientConfig;
import io.delopay.sdk.model.ProviderListResponse;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import static org.junit.jupiter.api.Assertions.*;

class ProvidersClientTest {
    private MockWebServer server;
    private DelopayClient client;

    @BeforeEach
    void setUp() throws IOException {
        server = new MockWebServer();
        server.start();
        client = new DelopayClient(new DelopayClientOptions(
                "test_api_key",
                server.url("/").toString(),
                10_000,
                0
        ));
    }

    @AfterEach
    void tearDown() throws IOException {
        server.shutdown();
    }

    @Test
    void listAllProviders() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody("{" +
                        "\"providers\":[" +
                        "{\"id\":\"STRIPE\",\"name\":\"Stripe\",\"enabled\":true," +
                        "\"supportedCurrencies\":[\"EUR\",\"USD\",\"GBP\"]," +
                        "\"features\":[\"cards\",\"sepa\",\"wallet\"]," +
                        "\"supportedCrypto\":[]}," +
                        "{\"id\":\"PAYPAL\",\"name\":\"PayPal\",\"enabled\":true," +
                        "\"supportedCurrencies\":[\"EUR\",\"USD\"]," +
                        "\"features\":[\"wallet\",\"paylater\"]," +
                        "\"supportedCrypto\":[]}," +
                        "{\"id\":\"NOWPAYMENTS\",\"name\":\"NOWPayments\",\"enabled\":true," +
                        "\"supportedCurrencies\":[\"USD\"]," +
                        "\"features\":[\"crypto\"]," +
                        "\"supportedCrypto\":[\"BTC\",\"ETH\",\"USDT\"]}" +
                        "]}"));

        ProviderListResponse response = client.providers().list();

        assertNotNull(response);
        assertEquals(3, response.providers().size());
        assertEquals("STRIPE", response.providers().get(0).id());
        assertTrue(response.providers().get(0).supportedCurrencies().contains("EUR"));
        assertTrue(response.providers().get(2).supportedCrypto().contains("BTC"));

        RecordedRequest request = server.takeRequest();
        assertEquals("/api/providers", request.getPath());
    }

    @Test
    void listEmptyProviders() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody("{\"providers\":[]}"));

        ProviderListResponse response = client.providers().list();

        assertNotNull(response);
        assertTrue(response.providers().isEmpty());
    }

    @Test
    void getStripeClientConfig() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody("{" +
                        "\"provider\":\"STRIPE\"," +
                        "\"publishableKey\":\"pk_test_1234567890\"," +
                        "\"clientId\":null}"));

        ProviderClientConfig response = client.providers().getClientConfig("stripe");

        assertNotNull(response);
        assertEquals("STRIPE", response.provider());
        assertEquals("pk_test_1234567890", response.publishableKey());

        RecordedRequest request = server.takeRequest();
        assertEquals("/api/providers/stripe/client-config", request.getPath());
    }

    @Test
    void getPayPalClientConfig() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody("{" +
                        "\"provider\":\"PAYPAL\"," +
                        "\"publishableKey\":null," +
                        "\"clientId\":\"paypal_client_id_123\"}"));

        ProviderClientConfig response = client.providers().getClientConfig("paypal");

        assertEquals("PAYPAL", response.provider());
        assertEquals("paypal_client_id_123", response.clientId());
        assertNull(response.publishableKey());
    }

    @Test
    void getClientConfigUrlEncoding() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody("{\"provider\":\"TEST\"}"));

        client.providers().getClientConfig("provider/test#1");

        RecordedRequest request = server.takeRequest();
        assertEquals("/api/providers/provider%2Ftest%231/client-config", request.getPath());
    }

    @Test
    void getClientConfigUnknownProvider() {
        server.enqueue(new MockResponse()
                .setResponseCode(400)
                .setBody("{\"message\":\"Unknown provider\",\"code\":\"E_UNKNOWN_PROVIDER\"}"));

        ApiError error = assertThrows(ApiError.class, () -> client.providers().getClientConfig("unknown"));
        assertEquals(400, error.status());
        assertEquals("E_UNKNOWN_PROVIDER", error.code());
    }

    @Test
    void getStripePaymentMethodsEurGermany() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody("{" +
                        "\"success\":true," +
                        "\"merchantCountry\":\"DE\"," +
                        "\"customerCountry\":\"DE\"," +
                        "\"currency\":\"EUR\"," +
                        "\"paymentMethods\":[" +
                        "{\"type\":\"card\",\"name\":\"Credit Card\",\"icon\":\"card.png\"}," +
                        "{\"type\":\"sepa_debit\",\"name\":\"SEPA Direct Debit\",\"icon\":\"sepa.png\"}," +
                        "{\"type\":\"ideal\",\"name\":\"iDEAL\",\"icon\":\"ideal.png\"}," +
                        "{\"type\":\"giropay\",\"name\":\"giropay\",\"icon\":\"giropay.png\"}" +
                        "]}"));

        PaymentMethodsResponse response = client.providers().getStripePaymentMethods("DE", "DE", "EUR");

        assertNotNull(response);
        assertTrue(response.success());
        assertEquals(4, response.paymentMethods().size());

        RecordedRequest request = server.takeRequest();
        String path = request.getPath();
        assertTrue(path.contains("merchantCountry=DE"));
        assertTrue(path.contains("customerCountry=DE"));
        assertTrue(path.contains("currency=EUR"));
    }

    @Test
    void getStripePaymentMethodsWithoutCurrency() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody("{" +
                        "\"success\":true," +
                        "\"merchantCountry\":\"US\"," +
                        "\"customerCountry\":\"US\"," +
                        "\"currency\":null," +
                        "\"paymentMethods\":[" +
                        "{\"type\":\"card\",\"name\":\"Credit Card\",\"icon\":\"card.png\"}," +
                        "{\"type\":\"ach_debit\",\"name\":\"ACH Direct Debit\",\"icon\":\"ach.png\"}" +
                        "]}"));

        PaymentMethodsResponse response = client.providers().getStripePaymentMethods("US", "US", null);

        assertEquals(2, response.paymentMethods().size());

        RecordedRequest request = server.takeRequest();
        String path = request.getPath();
        assertTrue(path.contains("merchantCountry=US"));
        assertTrue(path.contains("customerCountry=US"));
        assertFalse(path.contains("currency"));
    }

    @Test
    void getStripePaymentMethodsInvalidCountry() {
        server.enqueue(new MockResponse()
                .setResponseCode(400)
                .setBody("{\"message\":\"Invalid country code\",\"code\":\"E_INVALID_COUNTRY\"}"));

        ApiError error = assertThrows(ApiError.class, 
                () -> client.providers().getStripePaymentMethods("XX", "DE", null));
        assertEquals(400, error.status());
        assertEquals("E_INVALID_COUNTRY", error.code());
    }
}

File: sdks/typescript/tests/providers-client.test.ts
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { DelopayClient } from "../src";

function createMockResponse(body: object, status = 200, headers?: Record<string, string>): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: {
      "content-type": "application/json",
      ...headers
    }
  });
}

describe("ProvidersClient", () => {
  const fetchMock = vi.fn();
  let client: DelopayClient;

  beforeEach(() => {
    fetchMock.mockReset();
    vi.stubGlobal("fetch", fetchMock);
    client = new DelopayClient({
      apiKey: "test_api_key",
      baseUrl: "https://api.test.com"
    });
  });

  afterEach(() => {
    vi.unstubAllGlobals();
  });

  describe("list", () => {
    it("should list all available providers", async () => {
      const responseBody = {
        providers: [
          {
            id: "STRIPE",
            name: "Stripe",
            enabled: true,
            supportedCurrencies: ["EUR", "USD", "GBP"],
            features: ["cards", "sepa", "wallet"],
            supportedCrypto: []
          },
          {
            id: "PAYPAL",
            name: "PayPal",
            enabled: true,
            supportedCurrencies: ["EUR", "USD"],
            features: ["wallet", "paylater"],
            supportedCrypto: []
          },
          {
            id: "NOWPAYMENTS",
            name: "NOWPayments",
            enabled: true,
            supportedCurrencies: ["USD"],
            features: ["crypto"],
            supportedCrypto: ["BTC", "ETH", "USDT"]
          }
        ]
      };
      fetchMock.mockResolvedValue(createMockResponse(responseBody));

      const result = await client.providers.list();

      expect(fetchMock).toHaveBeenCalledTimes(1);
      const [url] = fetchMock.mock.calls[0] as [string, RequestInit];
      expect(url).toBe("https://api.test.com/api/providers");
      
      expect(result.providers).toHaveLength(3);
      expect(result.providers[0].id).toBe("STRIPE");
      expect(result.providers[0].supportedCurrencies).toContain("EUR");
      expect(result.providers[2].supportedCrypto).toContain("BTC");
    });

    it("should handle empty providers list", async () => {
      fetchMock.mockResolvedValue(createMockResponse({ providers: [] }));

      const result = await client.providers.list();
      expect(result.providers).toHaveLength(0);
    });
  });

  describe("getClientConfig", () => {
    it("should get Stripe client config", async () => {
      const responseBody = {
        provider: "STRIPE",
        publishableKey: "demo_publishable_token",
        clientId: null
      };
      fetchMock.mockResolvedValue(createMockResponse(responseBody));

      const result = await client.providers.getClientConfig("stripe");

      expect(fetchMock).toHaveBeenCalledTimes(1);
      const [url] = fetchMock.mock.calls[0] as [string, RequestInit];
      expect(url).toBe("https://api.test.com/api/providers/stripe/client-config");
      
      expect(result.provider).toBe("STRIPE");
      expect(result.publishableKey).toBe("demo_publishable_token");
    });

    it("should get PayPal client config", async () => {
      const responseBody = {
        provider: "PAYPAL",
        publishableKey: null,
        clientId: "paypal_client_id_123"
      };
      fetchMock.mockResolvedValue(createMockResponse(responseBody));

      const result = await client.providers.getClientConfig("paypal");

      expect(result.provider).toBe("PAYPAL");
      expect(result.clientId).toBe("paypal_client_id_123");
      expect(result.publishableKey).toBeNull();
    });

    it("should URL-encode provider ID", async () => {
      fetchMock.mockResolvedValue(createMockResponse({ provider: "TEST" }));

      await client.providers.getClientConfig("provider/test#1");

      const [url] = fetchMock.mock.calls[0] as [string, RequestInit];
      expect(url).toBe("https://api.test.com/api/providers/provider%2Ftest%231/client-config");
    });

    it("should handle unknown provider", async () => {
      fetchMock.mockResolvedValue(
        createMockResponse(
          { message: "Unknown provider", code: "E_UNKNOWN_PROVIDER" },
          400
        )
      );

      await expect(client.providers.getClientConfig("unknown")).rejects.toMatchObject({
        status: 400,
        code: "E_UNKNOWN_PROVIDER"
      });
    });
  });

  describe("getStripePaymentMethods", () => {
    it("should get payment methods for EUR in Germany", async () => {
      const responseBody = {
        success: true,
        merchantCountry: "DE",
        customerCountry: "DE",
        currency: "EUR",
        paymentMethods: [
          { type: "card", name: "Credit Card", icon: "card.png" },
          { type: "sepa_debit", name: "SEPA Direct Debit", icon: "sepa.png" },
          { type: "ideal", name: "iDEAL", icon: "ideal.png" },
          { type: "giropay", name: "giropay", icon: "giropay.png" }
        ]
      };
      fetchMock.mockResolvedValue(createMockResponse(responseBody));

      const result = await client.providers.getStripePaymentMethods({
        merchantCountry: "DE",
        customerCountry: "DE",
        currency: "EUR"
      });

      expect(fetchMock).toHaveBeenCalledTimes(1);
      const [url] = fetchMock.mock.calls[0] as [string, RequestInit];
      expect(url).toContain("/api/providers/stripe/payment-methods");
      expect(url).toContain("merchantCountry=DE");
      expect(url).toContain("customerCountry=DE");
      expect(url).toContain("currency=EUR");
      
      expect(result.success).toBe(true);
      expect(result.paymentMethods).toHaveLength(4);
      expect(result.paymentMethods.map(m => m.type)).toContain("ideal");
    });

    it("should get payment methods without currency parameter", async () => {
      const responseBody = {
        success: true,
        merchantCountry: "US",
        customerCountry: "US",
        currency: null,
        paymentMethods: [
          { type: "card", name: "Credit Card", icon: "card.png" },
          { type: "ach_debit", name: "ACH Direct Debit", icon: "ach.png" }
        ]
      };
      fetchMock.mockResolvedValue(createMockResponse(responseBody));

      const result = await client.providers.getStripePaymentMethods({
        merchantCountry: "US",
        customerCountry: "US"
      });

      const [url] = fetchMock.mock.calls[0] as [string, RequestInit];
      expect(url).toContain("merchantCountry=US");
      expect(url).toContain("customerCountry=US");
      expect(url).not.toContain("currency");
      
      expect(result.paymentMethods).toHaveLength(2);
    });

    it("should handle invalid country code", async () => {
      fetchMock.mockResolvedValue(
        createMockResponse(
          { message: "Invalid country code", code: "E_INVALID_COUNTRY" },
          400
        )
      );

      await expect(
        client.providers.getStripePaymentMethods({
          merchantCountry: "XX",
          customerCountry: "DE"
        })
      ).rejects.toMatchObject({
        status: 400,
        code: "E_INVALID_COUNTRY"
      });
    });
  });
});
